client/asset/eth/eth.go

```diff
@@ -2099,7 +2101,29 @@ func (w *TokenWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
 		return math.MaxUint64
 	}
 	return g.Transfer * feeRate
+}
+
+// RestorationInfo returns information about how to restore the wallet in
+// various external wallets.
+func (w *assetWallet) RestorationInfo(seed []byte) ([]*asset.WalletRestoration, error) {
+	extKey, err := keygen.GenDeepChild(seed, seedDerivationPath)
+	defer extKey.Zero()
+	if err != nil {
+		return nil, err
+	}
+	privateKey, err := extKey.SerializedPrivKey()
+	defer encode.ClearBytes(privateKey)
+	if err != nil {
+		return nil, err
+	}
 
+	return []*asset.WalletRestoration{
+		&asset.WalletRestoration{
+			Target:       "MetaMask",
+			Seed:         hex.EncodeToString(privateKey),
```

Comment:
I don't know what terminology is better to use, but not really a seed. It's an address's private key. Is this what it will be called when using in metamask?

Checking out metamask on firefox browser, it looks like it wants a word phrase:
![image](https://user-images.githubusercontent.com/34998433/172311771-fe2feac0-77da-4273-8382-fa618c102b7c.png)

Should we convert to words? We would rather have the user copy the words with pencil and paper than save the hex or copy paste it, I think.

Reply:
Yes, typing the private key in that box works. I should update the instructions to say that the wallet already needs to be initialized before the account can be added.

Going from seed phrase to seed is a one way operation. The link you shared was for generating the mnemonic, but the process for deriving the seed is below that:
https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed

Reply:
ah, right, because BIP39 mnemonic is encoded represenatation for "initial entropy" (not seed),

and, looks like having word-based seed requires some rework of `seed` concept (and related `PrimaryCredentials` structure) and its probably difficult to implement it in a backward-compatible manner ...

Reply:
> DEX requires some rework of seed concept

Do you mean the app seed?

Since this seed is something new and not used by anything but eth, we could do whatever we want with this "seed" I think, unless I'm missing something.

But the hex is fine with me. It's an emergency feature for refunds correct? Probably not used regularly.

We can't convert bytes -> mnemonic?

Reply:
> Do you mean the app seed?

Yes, I believe it is called `dex app seed`, this seed stored in this variable - https://github.com/decred/dcrdex/blob/a1d4690e5837c5882a26ff31034a657987bb00ba/client/db/types.go#L69-L71

let me try to lay out what happens next (maybe for future reference, pls correct me if I get it wrong)

`dex app seed` gets generated [here](https://github.com/decred/dcrdex/blob/a1d4690e5837c5882a26ff31034a657987bb00ba/client/core/core.go#L3373-L3380) and then **asset-specific seed** is derived from it as one-way operation [like this](https://github.com/decred/dcrdex/blob/a1d4690e5837c5882a26ff31034a657987bb00ba/client/core/core.go#L1872-L1880),
then **asset-specific seed** is used as `BIP 32 master-seed` when [deriving eth private key from it](https://github.com/decred/dcrdex/blob/a1d4690e5837c5882a26ff31034a657987bb00ba/client/asset/eth/eth.go#L336-L346) at `m/44'/60'/0'/0/0` path (and this **eth private key** gets exported by functionality implemented in this PR.

Now, `BIP 39 seed` would be the equivalent of DEX **asset-specific seed**, 
the problem is one can't convert `BIP 39 seed` (and hence DEX **asset-specific seed**) back into words, because it is [a one-way operation](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed).

Reply:
assuming I've layed out all bits correctly,

its probably possible to use `asset-specific seed` not as `BIP 39 seed` but as 
[BIP 39 entropy](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic),

and then we'll be able to recover this entropy in `RestorationInfo` func, and construct 24 word phrase from it,

I'm not suggesting that's must be done, just laying out possible options (I also don't have much context on why this feature gets implemented, but assume its nice to be able to take your DEX-generated wallet-phrase, for example, and use it in another app for whatever reason, better UI etc)

Reply:
That's the flow, yes.  This is almost entirely for recovery IMO.  We don't really want people messing in the wallet.
This is precisely why I made the stanalone assetseed app: https://github.com/decred/dcrdex/blob/master/client/cmd/assetseed/main.go

Because if all you have is your dex app seed and you want to access your funds without fixing dexc (presumably dexc is broken and you cannot use the feature added in this PR!), you just run the app and it gives your seed. In this case, I'd modify assetseed to take one more step for eth (deriving the private key from the asset seed since it's not a mneumonic that is recognized anywhere).

I don't really care much for all this generalization and expansion of the interfaces.  I'm more than happy with a recovery tool with asset specific code.

Reply:
> and, looks like having word-based seed in DEX requires some rework of `seed` concept (and related `PrimaryCredentials` structure) and its probably difficult to implement it in a backward-compatible manner

It's easy if we don't bother with BIP39 for the app seed <-> mneumonic at the top level.
https://github.com/decred/dcrwallet/blob/d1343d034a6431d90dc2500bc2d52a9ef4e9e029/walletseed/seed.go#L55
We can just adopt the same pgpwordlist and use ByteToMnemonic to make a mneumonic from our current app seed bytes.  This is unrelated to the metamask seed or other assets that use bip39 or other mneumonic schemes of course.

Reply:
Since app seed is not meant to be imported in any other app besides DEX I think it doesn't make much sense to make it BIP39-compliant,
representing app seed as words (from pgpwordlist linked above) in UI is preferable IMO (possibly, with an option to show it as binary seed),

it you agree that it's better UX, I can try looking into it and sketching up something working

Reply:
For the app seed, I think it would be somewhat of an improvement to show words using our (decred) scheme as we've discussed.  For making a paper wallet it's harder to mess up at least.

Reply:
Regarding the suggestion to use the `createWalletParams.Seed` as entropy to BIP39 mnemonic and seed generation, we could consider doing that even if we only give the account private key as in this PR.  Just to leave the door opened...

https://go.dev/play/p/vaMC6nA4k4P

https://iancoleman.io/bip39/
- Check the box "Show entropy details"
- Paste "f32a6d5737296e7d0dda7b78dfe5cf139a000374f7e32c886a19bfe183dcbe1b" into "Entropy" text box
- Select "12 Words" for "Mnemonic Length"
- Verify "BIP39 Mnemonic" and "BIP39 Seed" match the playground

I just don't know if that would be useful if other wallets do it a bit differently or use a password with the salt.

This obviously changes the account keys derived from a given app seed.

Reply:
Just want to point out one more time that taking the `m/44'/60'/0'/0/0` derivation steps to get an account private key doesn't have much point if there's no usable seed words corresponding to `m`.  This is not related to the app seed discussion, to be clear.

We just want to go from `createWalletParams.Seed` -> account private key, which we can get with `m/0'/0` for example.

My [previous comment](https://github.com/decred/dcrdex/pull/1648#discussion_r898428136) would go the other direction and insert another step so there is a seed to go with the `m/44'/60'/0'/0` path, just in case in the future we'd like to obtain a seed phrase that will derive the account private key we are exporting now.  I still prefer just exporting the private key though!

Am I missing something or is the `seedDerivationPath` used just because it's familiar?

Reply:
I think just because it's familiar.

Working on some Trevor stuff which also involved just needing an arbitrary, deterministic private key, the argument was made that starting with the seed and doing bip 39 derivation made it easier to derive addresses with other software, if you ever needed to.

Yes we only use the one address atm, but maybe in the future there's more we want to do here.

> no usable seed words corresponding to m.

I guess I don't understand what this means. We COULD export `m` in the future?

Reply:
Oh, we would need "seed words" for most wallets in order to import the seed? because we cannot go from seed hex -> words apparently?huh...

Reply:
Right.  I get the reasoning of making it "easier to derive addresses with other software, if you ever needed to", but that means words, not the seed bytes that come from words, which are presently non-existent.  Hence the playground link I posted, if we want to have yet another dep "just in case" we want words available in the future for this purpose.  I dunno

Reply:
I think it's a good idea to be able to generate the words if needed. I can do a separate PR for that though, that doesn't really affect anything in this one.